examples/app-pytorch/app_pytorch/fedavg.py

```diff
@@ -0,0 +1,390 @@
+        if weight is None:
+            log(
+                WARN,
+                "No weighting key '%s' found in MetricRecord, defaulting to 1.0 for all Messages",
+                weighting_metric_name,
+            )
+            log(
+                WARN,
+                "Ensure your ClientApps include in their replies a {'metrics': MetricRecord({'%s': ...})}",
+                weighting_metric_name,
+            )
+            total_examples = len(records)
+            examples_per_record = [1.0] * len(records)
+            break
+        else:
+            total_examples += weight
+            examples_per_record.append(weight)
+
+    # Perform weighted aggregation
+    aggregated_arrays = ArrayRecord()
+    aggregated_metrics = MetricRecord()
+    for record, weight in zip(records, examples_per_record):
+        for record_item in record.values():
+            # For ArrayRecord
+            if isinstance(record_item, ArrayRecord):
+                # aggregate in-place
+                # TODO: optimize so at least we don't need to serde all the time
+                for key, value in record_item.items():
+                    if key not in aggregated_arrays:
+                        aggregated_arrays[key] = Array(
+                            value.numpy() * weight / total_examples
+                        )
+                    else:
+                        aggregated_arrays[key] = Array(
+                            aggregated_arrays[key].numpy()
+                            + value.numpy() * weight / total_examples
+                        )
+
+            # For MetricRecord
+            elif isinstance(record_item, MetricRecord):
+                # aggregate in-place
+                for key, value in record_item.items():
+                    if key == weighting_metric_name:
+                        continue  # ! we don't want to keep this key as part of the aggregated recorddict, do we?
+                    if key not in aggregated_metrics:
+                        if isinstance(value, list):
+                            aggregated_metrics[key] = (
+                                np.array(value) * weight / total_examples
+                            ).tolist()
+                        else:
+                            aggregated_metrics[key] = value * weight / total_examples
+                    else:
+                        if isinstance(value, list):
+                            aggregated_metrics[key] = (
+                                np.array(aggregated_metrics[key])
+                                + np.array(value) * weight / total_examples
+                            ).tolist()
+                        else:
+                            aggregated_metrics[key] += value * weight / total_examples
+
+    return aggregated_arrays, aggregated_metrics
+
+
+def sample_nodes(
+    grid: Grid, min_sample_nodes: int, fraction_sample: float
+) -> tuple[list[int], list[int]]:
```

Comment:
```suggestion
def sample_nodes(
    grid: Grid, min_sample_nodes: int, fraction_sample: float
) -> list[int]:
```

This function looks a bit strange to me. Maybe we could make it more focused and intuitive, like returning the sampled node IDs? If users do need all node IDs, they can simply call `grid.get_node_ids()`.